File: lib/os_mon/c_src/cpu_sup.c
/* ``The contents of this file are subject to the Erlang Public License,
 * Version 1.1, (the "License"); you may not use this file except in
 * compliance with the License. You should have received a copy of the
 * Erlang Public License along with this software. If not, it can be
 * retrieved via the world wide web at http://www.erlang.org/.
 * 
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and limitations
 * under the License.
 * 
 * The Initial Developer of the Original Code is Ericsson Utvecklings AB.
 * Portions created by Ericsson are Copyright 1999, Ericsson Utvecklings
 * AB. All Rights Reserved.''
 * 
 *     $Id$
 */
/*
 * CPU supervision
 *
 * Uses kstat library only available on Solaris 2
 * Compile with: gcc -o cpu_sup cpu_sup.c -lkstat
 *
 * Use open_port({spawn,Prog},[stream]) to communicate.
 * 
 */

#include<stdio.h>
#include<kstat.h>

#define FD_IN    0
#define FD_OUT   1
#define FD_ERR   2

#define PING     'p'
#define NPROCS   'n'
#define AVG1     '1'
#define AVG5     '5'
#define AVG15    'f'
#define QUIT     'q'

/*#define DISPLAY_ON_STDERR*/

static unsigned int measure(char* name);
static void send(unsigned int data);
static void error(char* err_msg);

static kstat_ctl_t *kc;
static kstat_t *ks;

int main(int argc, char** argv) {
  kid_t kid;
  char cmd;
  int rc;

  kc = kstat_open();
  if(!kc) error("Can't open header kstat");

  ks = kstat_lookup(kc,"unix",0,"system_misc");
  if(!ks) error("Can't open system_misc kstat");


  while(1) {
    rc = read(FD_IN, &cmd, sizeof(cmd));
    if(rc <= 0)
      error(rc<0?"Error reading from Erlang":"Erlang has closed");
    
    switch(cmd) {
    case PING:   send(4711); break;
    case NPROCS: send(measure("nproc")); break;
    case AVG1:   send(measure("avenrun_1min")); break;
    case AVG5:   send(measure("avenrun_5min")); break;
    case AVG15:  send(measure("avenrun_15min")); break;
    case QUIT:   return 0;
    default:     error("Bad command"); break;
    }
  }
  return 0; /* supress warnings */
}

static unsigned int measure(char* name) {
  kstat_named_t* entry;

  kstat_read(kc,ks,NULL);
  entry = kstat_data_lookup(ks,name);
  if(!entry)
    return -1;
  
  if(entry->data_type != KSTAT_DATA_ULONG)
    return -1;

  return entry->value.ul;
}

static void send(unsigned int data) {
  char buf[4];
  int rc;

  buf[0] = (data >> 24) & 0xff;
  buf[1] = (data >> 16) & 0xff;
  buf[2] = (data >>  8) & 0xff;
  buf[3] = (data      ) & 0xff;

  rc = write(FD_OUT, buf, sizeof(buf));

  if(rc != sizeof(buf))
    error("Error writing to Erlang");

#ifdef DISPLAY_ON_STDERR
  { /* Debugging aid */
    char text[80];
    sprintf(text,"%d (%d,%d,%d,%d)\n",data,buf[0],buf[1],buf[2],buf[3]);
    rc = write(FD_ERR,text,strlen(text));
  }
#endif
}

static void error(char* err_msg) {
  write(FD_ERR, err_msg, strlen(err_msg));
  write(FD_ERR, "\n", 1);
  exit(-1);
}

File: lib/os_mon/src/cpu_sup.erl
%% ``The contents of this file are subject to the Erlang Public License,
%% Version 1.1, (the "License"); you may not use this file except in
%% compliance with the License. You should have received a copy of the
%% Erlang Public License along with this software. If not, it can be
%% retrieved via the world wide web at http://www.erlang.org/.
%% 
%% Software distributed under the License is distributed on an "AS IS"
%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
%% the License for the specific language governing rights and limitations
%% under the License.
%% 
%% The Initial Developer of the Original Code is Ericsson Utvecklings AB.
%% Portions created by Ericsson are Copyright 1999, Ericsson Utvecklings
%% AB. All Rights Reserved.''
%% 
%%     $Id$
%%
-module(cpu_sup).

%%% Purpose : Obtain cpu statistics

-export([nprocs/0,avg1/0,avg5/0,avg15/0,ping/0]).

%% External exports
-export([start_link/0, start/0, stop/0]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).

-define(APPLICATION,"os_mon").
-define(PORT_PROG,"/bin/cpu_sup"). %% This is relative priv_dir
-define(NAME,cpu_sup).

%% Internal protocol with the port program
-define(nprocs,"n").
-define(avg1,"1").
-define(avg5,"5").
-define(avg15,"f").
-define(quit,"q").
-define(ping,"p").

-record(state, {port}).

%%%----------------------------------------------------------------------
%%% API
%%%----------------------------------------------------------------------

start()  -> gen_server:start({local, cpu_sup}, cpu_sup, [], []).
start_link() -> gen_server:start_link({local, cpu_sup}, cpu_sup, [], []).
stop()   -> gen_server:call(?NAME,?quit).

nprocs() -> gen_server:call(?NAME,?nprocs).
avg1()   -> gen_server:call(?NAME,?avg1).
avg5()   -> gen_server:call(?NAME,?avg5).
avg15()  -> gen_server:call(?NAME,?avg15).
ping()   -> gen_server:call(?NAME,?ping).

%%%----------------------------------------------------------------------
%%% Callback functions from gen_server
%%%----------------------------------------------------------------------

%%----------------------------------------------------------------------
%% Func: init/1
%% Returns: {ok, State}          |
%%          {ok, State, Timeout} |
%%          {stop, Reason}
%%----------------------------------------------------------------------
init([]) ->
    case os:type() of
	{unix,sunos} ->
	    Prog = code:priv_dir(?APPLICATION) ++ ?PORT_PROG,
	    Port = open_port({spawn,Prog},[stream]),
	    if port(Port) ->
		    {ok, #state{port=Port}};
	       true ->
		    {stop, {port_prog_not_available,Port}}
	    end;
	{unix,linux} ->
	    {ok,#state{port=not_used}};
	{unix,freebsd} ->
	    {ok,#state{port=not_used}};
	{unix,openbsd} ->
	    {ok,#state{port=not_used}}
    end.

%%----------------------------------------------------------------------
%% Func: handle_call/3
%% Returns: {reply, Reply, State}          |
%%          {reply, Reply, State, Timeout} |
%%          {noreply, State}               |
%%          {noreply, State, Timeout}      |
%%          {stop, Reason, Reply, State}   | (terminate/2 is called)
%%          {stop, Reason, Reply, State}     (terminate/2 is called)
%%----------------------------------------------------------------------
handle_call(?quit, From, State) ->
    case State#state.port of
	not_used ->
	    ok;
	Port ->
	    Port ! {self(), {command, ?quit}},
	    Port ! {self(), close}
    end,
    {stop, shutdown, ok, State};
handle_call(Request, From, State) ->
    Reply = get_measurement(Request,State#state.port),
    {reply, Reply, State}.

%%----------------------------------------------------------------------
%% Func: handle_cast/2
%% Returns: {noreply, State}          |
%%          {noreply, State, Timeout} |
%%          {stop, Reason, State}            (terminate/2 is called)
%%----------------------------------------------------------------------
handle_cast(Msg, State) ->
    {noreply, State}.

%%----------------------------------------------------------------------
%% Func: handle_info/2
%% Returns: {noreply, State}          |
%%          {noreply, State, Timeout} |
%%          {stop, Reason, State}            (terminate/2 is called)
%%----------------------------------------------------------------------
handle_info({Port,closed}, State) when Port == State#state.port ->
    {stop, port_closed, State#state{port=closed}};
handle_info(Info, State) ->
    {noreply, State}.

%%----------------------------------------------------------------------
%% Func: terminate/2
%% Purpose: Shutdown the server
%% Returns: any (ignored by gen_server)
%%----------------------------------------------------------------------
terminate(Reason, State) ->
    ok.

%%%----------------------------------------------------------------------
%%% Internal functions
%%%----------------------------------------------------------------------

get_measurement(Request,Port) ->
    case os:type() of
	{unix,sunos} ->
	    Port ! {self(), {command, Request}},
	    receive
		{Port,{data,[D3,D2,D1,D0]}} ->
		    (D3 bsl 24) bor (D2 bsl 16) bor (D1 bsl 8) bor D0
	    end;
	{unix,linux} ->
	    {ok,F} = file:open("/proc/loadavg",[read,raw]),
	    {ok,D} = file:read(F,24),
	    ok = file:close(F),
	    {ok,[Load1,Load5,Load15,PRun,PTotal],_} =
		io_lib:fread("~f ~f ~f ~d/~d", D),
	    case Request of
		?avg1  -> sunify(Load1);
		?avg5  -> sunify(Load5);
		?avg15 -> sunify(Load15);
		?ping -> 4711;
		?nprocs -> PTotal
	    end;
	{unix,freebsd} ->
	    D = os:cmd("/sbin/sysctl -n vm.loadavg") -- "\n",
	    {ok,[Load1,Load5,Load15],_} =
		io_lib:fread("{ ~f ~f ~f }", D),
	    % We could count the lines from the ps command as well
	    case Request of
		?avg1  -> sunify(Load1);
		?avg5  -> sunify(Load5);
		?avg15 -> sunify(Load15);
		?ping -> 4711;
		?nprocs ->
		    {ok, DirList} = file:list_dir("/proc"),
		    length(DirList)
	    end;
	{unix,openbsd} ->
	    D = os:cmd("/sbin/sysctl -n vm.loadavg") -- "\n",
	    {ok, [L1, L5, L15], _} = io_lib:fread("~f ~f ~f", D),
	    case Request of
		?avg1  -> sunify(L1);
		?avg5  -> sunify(L5);
		?avg15 -> sunify(L15);
		?ping -> 4711;
		?nprocs ->
		    Ps = os:cmd("/bin/ps -ax | /usr/bin/wc -l"),
		    {ok, [N], _} = io_lib:fread("~d", Ps),
		    N-1
	    end
    end.

sunify(Val)  ->
    round(Val*256). % Note that Solaris and Linux load averages are
		    % measured quite differently anyway

%%%----------------------------------------------------------------------

